AuthenticationExpired errors when connecting to wifi

## Bug description

<!-- A clear and concise description of what the bug is. -->

When using esp-radio to connect to a network i'm constantly getting AuthenticationExpired errors. Which for the life of me i cant figure out. I've tried my home network and my iPhone's hotspot with maximum compatibility turned on. Both give the same errors. 

I've tried to turn off the password on my home network to check if we can bypass the authentication but also here it fails. 

Hardware seems to be ok as i'm able to connect to my wifi when flashing with esphome software.

## To Reproduce

Im using the following code to conect to my network:

```rust
use embassy_net::{Config as NetConfig, DhcpConfig, Runner, Stack, StackResources};
use embassy_time::{Duration, Timer};
use esp_radio::wifi::{Config, Interface, PowerSaveMode, WifiController,
    sta::{ScanMethod, StationConfig},
    ControllerConfig,
};
use static_cell::StaticCell;

pub struct Wifi {
    pub stack: Stack<'static>,
}

impl Wifi {
    pub fn new(
        wifi_peripheral: esp_hal::peripherals::WIFI<'static>,
        spawner: &embassy_executor::Spawner,
    ) -> Self {
        let station_config = Config::Station(
            StationConfig::default()
                .with_ssid(env!("WIFI_SSID"))
                .with_password(env!("WIFI_PASSWORD").into())
        );

        let (wifi_controller, interfaces) = esp_radio::wifi::new(
            wifi_peripheral,
            ControllerConfig::default().with_initial_config(station_config),
        )
        .expect("Failed to create Wi-Fi controller");

        static WIFI_CONTROLLER: StaticCell<WifiController<'static>> = StaticCell::new();
        let wifi_controller = WIFI_CONTROLLER.init(wifi_controller);

        let net_config = NetConfig::dhcpv4(DhcpConfig::default());

        static RESOURCES: StaticCell<StackResources<5>> = StaticCell::new();
        let resources = RESOURCES.init(StackResources::new());

        let net_seed = esp_hal::rng::Rng::new().random() as u64;
        let (stack, runner) =
            embassy_net::new(interfaces.station, net_config, resources, net_seed);

        spawner.spawn(connection_task(wifi_controller).expect("Could not spawn connection_task"));
        spawner.spawn(net_task(runner).expect("Could not spawn net_task"));

        Self { stack }
    }
}

#[embassy_executor::task]
async fn connection_task(controller: &'static mut WifiController<'static>) {
    loop {
        if !controller.is_connected() {
            log::info!("Connecting...");
            match controller.connect_async().await {
                Ok(info) => log::info!("Connected! {:?}", info),
                Err(e) => {
                    log::error!("Connect failed: {:?}", e);
                    Timer::after(Duration::from_secs(5)).await;
                }
            }
        } else {
            match controller.wait_for_disconnect_async().await {
                Ok(info) => log::info!("Disconnected: {:?}", info),
                Err(e) => log::warn!("Disconnect wait error: {:?}", e),
            }
            Timer::after(Duration::from_secs(5)).await;
        }
    }
}

#[embassy_executor::task]
async fn net_task(mut runner: Runner<'static, Interface<'static>>) {
    runner.run().await;
}
```

This returns with the following error:

```bash
ERROR - Connect failed: Disconnected(DisconnectedStationInfo { ssid: "iPhone", bssid: [74, 221, 65, 19, 204, 151], reason: AuthenticationExpired, rssi: -37 })
```

Also see below for full driver log prints

<!-- Please share a minimal reproducer which demonstrates the issue. `documentation/REPRODUCERS.md` gives some hints what an ideal reproducer should look like. -->

<!-- Make sure you are able to reproduce the bug in the `main` branch, too. -->

## Expected behavior

<!-- A clear and concise description of what you expected to happen. Attach screenshots if needed. -->

I expect it to connect to my network. 


## Environment

- Target device: espflash board-info
[2026-05-30T13:27:02Z INFO ] Serial port: '/dev/cu.usbmodem112301'
[2026-05-30T13:27:02Z INFO ] Connecting...
[2026-05-30T13:27:02Z INFO ] Using flash stub
Chip type:         esp32s3 (revision v0.2)
Crystal frequency: 40 MHz
Flash size:        16MB
Features:          WiFi, BLE, Embedded Flash
MAC address:       14:c1:9f:d7:dc:a4
Flags: 0x00000000 (0)
Key Purposes: [0, 0, 0, 0, 0, 0, 12]
Chip ID: 9
API Version: 0
Secure Boot: Disabled
Flash Encryption: Disabled
SPI Boot Crypt Count (SPI_BOOT_CRYPT_CNT): 0x0
 <!-- Use `espflash board-info` to get the target device information. -->
- Crate name and version: esp-hal 1.1.0 and esp-radio 0.18.0

### Full log driver output

```
INFO - vendor id    : 0d (AP)
INFO - dev id       : 02 (generation 3)
INFO - density      : 03 (64 Mbit)
INFO - good-die     : 01 (Pass)
INFO - Latency      : 01 (Fixed)
INFO - VCC          : 01 (3V)
INFO - SRF          : 01 (Fast Refresh)
INFO - BurstType    : 01 (Hybrid Wrap)
INFO - BurstLen     : 01 (32 Byte)
INFO - Readlatency  : 02 (10 cycles@Fixed)
INFO - DriveStrength: 00 (1/1)
INFO - 8388608 bytes of PSRAM
INFO - PSRAM initialized successfully in Octal SPI mode
INFO - ESP_RADIO pp rom version: e7ae62f
INFO - ESP_RADIO net80211 rom version: e7ae62f
INFO - wifi I (318) wifi:
INFO - wifi wifi driver task: 3fcddcf0, prio:29, stack:6656, core=0
INFO - wifi I (326) wifi:
INFO - wifi wifi firmware version: 4df78f2
INFO - wifi I (331) wifi:
INFO - wifi wifi certification version: v7.0
INFO - wifi I (338) wifi:
INFO - wifi config NVS flash: disabled
INFO - wifi I (344) wifi:
INFO - wifi config nano formatting: disabled
INFO - wifi I (350) wifi:
INFO - wifi Init data frame dynamic rx buffer num: 32
INFO - wifi I (358) wifi:
INFO - wifi Init static rx mgmt buffer num: 5
INFO - wifi I (365) wifi:
INFO - wifi Init management short buffer num: 32
INFO - wifi I (372) wifi:
INFO - wifi Init dynamic tx buffer num: 32
INFO - wifi I (378) wifi:
INFO - wifi Init static tx FG buffer num: 2
INFO - wifi I (385) wifi:
INFO - wifi Init static rx buffer size: 1600
INFO - wifi I (392) wifi:
INFO - wifi Init static rx buffer num: 10
INFO - wifi I (398) wifi:
INFO - wifi Init dynamic rx buffer num: 32
INFO - wifi D (405) wifi:
INFO - wifi Reset scan defaults in ms : min,max,passive,chan_dwell :(0,120,360,30)
INFO - wifi V (408) wifi:
INFO - wifi hint=255, act_dur=<0,120>, pas_dur=360, nchans=14 history_num=0
INFO - wifi I (418) wifi:
INFO - wifi (2.4G)set country: cc=CN
INFO - wifi I (423) wifi:
INFO - wifi Set ps type: 0, coexist: 0
INFO - wifi D (430) wifi:
INFO - wifi clear blacklist
INFO - wifi V (470) wifi:
INFO - wifi ht20 freq=2412, chan=1
INFO - wifi D (472) wifi:
INFO - wifi filter: set rx policy=0
INFO - wifi I (474) wifi:
INFO - wifi mode : sta (14:c1:9f:d7:dc:a4)
INFO - wifi I (477) wifi:
INFO - wifi enable tsf
INFO - wifi D (481) wifi:
INFO - wifi filter: set rx policy=1
INFO - wifi D (488) wifi:
INFO - wifi connect status 0 -> 0
INFO - wifi I (495) wifi:
INFO - wifi Set ps type: 0, coexist: 0
INFO - Connecting...
INFO - wifi D (500) wifi:
INFO - wifi Start wifi connect
INFO - wifi D (506) wifi:
INFO - wifi connect status 0 -> 0
INFO - wifi D (511) wifi:
INFO - wifi connect chan=0
INFO - wifi V (516) wifi:
INFO - wifi nvs=0, ssid=iPhone, channel=255
INFO - wifi V (522) wifi:
INFO - wifi ssid=iPhone match nvs 0, channel=255
INFO - wifi D (529) wifi:
INFO - wifi first chan=1
INFO - wifi D (534) wifi:
INFO - wifi connect status 0 -> 1
INFO - wifi D (540) wifi:
INFO - wifi filter: set rx policy=3
INFO - wifi D (546) wifi:
INFO - wifi clear scan ap list
INFO - wifi D (551) wifi:
INFO - wifi start scan: type=0x50f, priority=2, cb=42079618, arg=00000000, ss_state=0x1, t
INFO - wifi V (562) wifi:
INFO - wifi inter_channel_timeout: arg=00000000, ss_state=0x1
INFO - wifi D (570) wifi:
INFO - wifi perform scan: ss_state=0x9, chan<1,0>, dur<0,120>
INFO - wifi V (578) wifi:
INFO - wifi change: chan<1,0>, dur<0,120>
INFO - wifi V (584) wifi:
INFO - wifi enter start op, arg=3FC98DC0
INFO - wifi V (591) wifi:
INFO - wifi scan operation: state=0x3, chan<1,0>, arg=3FC98DC0, status=0
INFO - wifi V (600) wifi:
INFO - wifi scan specific ssid=iPhone
INFO - wifi V (606) wifi:
INFO - wifi start max timer
INFO - wifi D (731) wifi:
INFO - wifi scan end: arg=3FC98DC0, status=0, ss_state=0x3
INFO - wifi D (733) wifi:
INFO - wifi perform scan: ss_state=0x9, chan<2,0>, dur<0,120>
INFO - wifi V (737) wifi:
INFO - wifi change: chan<2,0>, dur<0,120>
INFO - wifi V (743) wifi:
INFO - wifi ht20 freq=2417, chan=2
INFO - wifi V (748) wifi:
INFO - wifi enter start op, arg=3FC98DC0
INFO - wifi V (754) wifi:
INFO - wifi scan operation: state=0x3, chan<2,0>, arg=3FC98DC0, status=0
INFO - wifi V (763) wifi:
INFO - wifi scan specific ssid=iPhone
INFO - wifi V (769) wifi:
INFO - wifi start max timer
INFO - wifi D (894) wifi:
INFO - wifi scan end: arg=3FC98DC0, status=0, ss_state=0x3
INFO - wifi D (895) wifi:
INFO - wifi perform scan: ss_state=0x9, chan<3,0>, dur<0,120>
INFO - wifi V (899) wifi:
INFO - wifi change: chan<3,0>, dur<0,120>
INFO - wifi V (906) wifi:
INFO - wifi ht20 freq=2422, chan=3
INFO - wifi V (912) wifi:
INFO - wifi enter start op, arg=3FC98DC0
INFO - wifi V (918) wifi:
INFO - wifi scan operation: state=0x3, chan<3,0>, arg=3FC98DC0, status=0
INFO - wifi V (927) wifi:
INFO - wifi scan specific ssid=iPhone
INFO - wifi V (931) wifi:
INFO - wifi start max timer
INFO - wifi D (1056) wifi:
INFO - wifi scan end: arg=3FC98DC0, status=0, ss_state=0x3
INFO - wifi D (1057) wifi:
INFO - wifi perform scan: ss_state=0x9, chan<4,0>, dur<0,120>
INFO - wifi V (1061) wifi:
INFO - wifi change: chan<4,0>, dur<0,120>
INFO - wifi V (1068) wifi:
INFO - wifi ht20 freq=2427, chan=4
INFO - wifi V (1074) wifi:
INFO - wifi enter start op, arg=3FC98DC0
INFO - wifi V (1081) wifi:
INFO - wifi scan operation: state=0x3, chan<4,0>, arg=3FC98DC0, status=0
INFO - wifi V (1090) wifi:
INFO - wifi scan specific ssid=iPhone
INFO - wifi V (1096) wifi:
INFO - wifi start max timer
INFO - wifi D (1221) wifi:
INFO - wifi scan end: arg=3FC98DC0, status=0, ss_state=0x3
INFO - wifi D (1222) wifi:
INFO - wifi perform scan: ss_state=0x9, chan<5,0>, dur<0,120>
INFO - wifi V (1227) wifi:
INFO - wifi change: chan<5,0>, dur<0,120>
INFO - wifi V (1233) wifi:
INFO - wifi ht20 freq=2432, chan=5
INFO - wifi V (1240) wifi:
INFO - wifi enter start op, arg=3FC98DC0
INFO - wifi V (1246) wifi:
INFO - wifi scan operation: state=0x3, chan<5,0>, arg=3FC98DC0, status=0
INFO - wifi V (1255) wifi:
INFO - wifi scan specific ssid=iPhone
INFO - wifi V (1261) wifi:
INFO - wifi start max timer
INFO - wifi D (1386) wifi:
INFO - wifi scan end: arg=3FC98DC0, status=0, ss_state=0x3
INFO - wifi D (1387) wifi:
INFO - wifi perform scan: ss_state=0x9, chan<6,0>, dur<0,120>
INFO - wifi V (1391) wifi:
INFO - wifi change: chan<6,0>, dur<0,120>
INFO - wifi V (1398) wifi:
INFO - wifi ht20 freq=2437, chan=6
INFO - wifi V (1404) wifi:
INFO - wifi enter start op, arg=3FC98DC0
INFO - wifi V (1409) wifi:
INFO - wifi scan operation: state=0x3, chan<6,0>, arg=3FC98DC0, status=0
INFO - wifi V (1418) wifi:
INFO - wifi scan specific ssid=iPhone
INFO - wifi V (1424) wifi:
INFO - wifi start max timer
INFO - wifi D (1429) wifi:
INFO - wifi rsn valid: gcipher=3 ucipher=3 akm=5 mac=ba:cd:60:65:0d:81
INFO - wifi V (1438) wifi:
INFO - wifi updated connection retries to 1
INFO - wifi D (1445) wifi:
INFO - wifi profile match: ss_state=0x7
INFO - wifi V (1451) wifi:
INFO - wifi scan histroy table is not full yet, add
INFO - wifi V (1459) wifi:
INFO - wifi add ssid=iPhone, chan=6 to scan history
INFO - wifi D (1466) wifi:
INFO - wifi scan end: arg=3FC98DC0, status=0, ss_state=0x7
INFO - wifi D (1473) wifi:
INFO - wifi find first mathched ssid, scan done
INFO - wifi V (1480) wifi:
INFO - wifi back home chan=<1,0>, current chan=<6,0>
INFO - wifi V (1487) wifi:
INFO - wifi ht20 freq=2412, chan=1
INFO - wifi D (1494) wifi:
INFO - wifi filter: set rx policy=4
INFO - wifi D (1499) wifi:
INFO - wifi first chan=1
INFO - wifi V (1504) wifi:
INFO - wifi scan_done: arg=00000000, status=0, cur_time=1504450, scan_id=128, scan state=0
INFO - wifi V (1514) wifi:
INFO - wifi call scan_done cb, arg=00000000
INFO - wifi D (1521) wifi:
INFO - wifi handoff_cb: status=0
INFO - wifi V (1526) wifi:
INFO - wifi best bss has set.
INFO - wifi D (1532) wifi:
INFO - wifi ap found, mac=ba:cd:60:65:0d:81
INFO - wifi V (1539) wifi:
INFO - wifi bssid=ba:cd:60:65:0d:81, LR=0
INFO - wifi D (2127) wifi:
INFO - wifi going for connection with bssid=ba:cd:60:65:0d:81
INFO - wifi D (2129) wifi:
INFO - wifi new_bss=3FC99590, cur_bss=00000000, new_chan=<6,0>, cur_chan=1
INFO - wifi D (2134) wifi:
INFO - wifi filter: set rx policy=5
INFO - wifi I (2140) wifi:
INFO - wifi new:<6,0>, old:<1,0>, ap:<255,255>, sta:<6,0>, prof:1, snd_ch_cfg:0x0
INFO - wifi V (2149) wifi:
INFO - wifi ht20 freq=2437, chan=6
INFO - wifi D (2155) wifi:
INFO - wifi connect_op: status=0, auth=5, cipher=3
INFO - wifi D (2162) wifi:
INFO - wifi auth mode is not none
INFO - wifi D (2168) wifi:
INFO - wifi connect_bss: auth=1, reconnect=0
INFO - wifi I (2175) wifi:
INFO - wifi state: init -> auth (0xb0)
INFO - wifi D (2182) wifi:
INFO - wifi start 1000ms AUTH timer
INFO - wifi D (2187) wifi:
INFO - wifi clear scan ap list
INFO - wifi D (3187) wifi:
INFO - wifi auth timeout
INFO - wifi I (3188) wifi:
INFO - wifi state: auth -> init (0x200)
INFO - wifi D (3190) wifi:
INFO - wifi stop beacon and connect timers
INFO - wifi D (3194) wifi:
INFO - wifi connect status 1 -> 4
INFO - wifi D (3200) wifi:
INFO - wifi reason: auth expire(2)
INFO - wifi D (3206) wifi:
INFO - wifi add bssid ba:cd:60:65:0d:81 to blacklist, cnt=0
INFO - wifi D (3213) wifi:
INFO - wifi stop CSA timer
INFO - wifi V (3218) wifi:
INFO - wifi bss connection retry cnt=1
INFO - wifi D (3225) wifi:
INFO - wifi remove bssid ba:cd:60:65:0d:81 from blacklist, cnt=1
INFO - wifi D (3233) wifi:
INFO - wifi connect status 4 -> 0
INFO - wifi I (3238) wifi:
INFO - wifi new:<6,0>, old:<6,0>, ap:<255,255>, sta:<6,0>, prof:1, snd_ch_cfg:0x0
INFO - wifi D (3246) wifi:
INFO - wifi filter: set rx policy=8
INFO - wifi D (3252) wifi:
INFO - wifi connect status 0 -> 1
INFO - wifi D (3258) wifi:
INFO - wifi Try to connect with next matched AP
INFO - wifi V (3264) wifi:
INFO - wifi best bss has set.
INFO - wifi D (3270) wifi:
INFO - wifi ap found, mac=ba:cd:60:65:0d:81
INFO - wifi V (3277) wifi:
INFO - wifi bssid=ba:cd:60:65:0d:81, LR=0
INFO - wifi D (3284) wifi:
INFO - wifi going for connection with bssid=ba:cd:60:65:0d:81
INFO - wifi D (3292) wifi:
INFO - wifi new_bss=3FC99590, cur_bss=00000000, new_chan=<6,0>, cur_chan=6
INFO - wifi D (3301) wifi:
INFO - wifi filter: set rx policy=5
INFO - wifi I (3307) wifi:
INFO - wifi new:<6,0>, old:<6,0>, ap:<255,255>, sta:<6,0>, prof:1, snd_ch_cfg:0x0
INFO - wifi D (3317) wifi:
INFO - wifi connect_op: status=0, auth=5, cipher=3
INFO - wifi D (3323) wifi:
INFO - wifi auth mode is not none
INFO - wifi D (3330) wifi:
INFO - wifi connect_bss: auth=1, reconnect=0
INFO - wifi I (3336) wifi:
INFO - wifi state: init -> auth (0xb0)
INFO - wifi D (3343) wifi:
INFO - wifi start 1000ms AUTH timer
INFO - wifi D (4349) wifi:
INFO - wifi auth timeout
INFO - wifi I (4350) wifi:
INFO - wifi state: auth -> init (0x200)
INFO - wifi D (4351) wifi:
INFO - wifi stop beacon and connect timers
INFO - wifi D (4356) wifi:
INFO - wifi connect status 1 -> 4
INFO - wifi D (4362) wifi:
INFO - wifi reason: auth expire(2)
INFO - wifi D (4368) wifi:
INFO - wifi add bssid ba:cd:60:65:0d:81 to blacklist, cnt=0
INFO - wifi D (4375) wifi:
INFO - wifi stop CSA timer
INFO - wifi V (4380) wifi:
INFO - wifi bss connection retry cnt=0
INFO - wifi D (4385) wifi:
INFO - wifi remove ba:cd:60:65:0d:81 from rc list
INFO - wifi D (4392) wifi:
INFO - wifi connect status 4 -> 0
INFO - wifi I (4397) wifi:
INFO - wifi new:<6,0>, old:<6,0>, ap:<255,255>, sta:<6,0>, prof:1, snd_ch_cfg:0x0
INFO - wifi D (4407) wifi:
INFO - wifi filter: set rx policy=8
INFO - wifi D (4413) wifi:
INFO - wifi Send disconnect event, reason=2, AP number=0
ERROR - Connect failed: Disconnected(DisconnectedStationInfo { ssid: "iPhone", bssid: [186, 205, 96, 101, 13, 129], reason: AuthenticationExpired, rssi: -52 })
```

Hi @MabezDev,

The issue persists with 1.0.0-beta.0.

I have [uploaded the code](https://github.com/VasanthakumarV/esp32c3-wifi-html-server/tree/main) I ran _(the [wifi code](https://github.com/VasanthakumarV/esp32c3-wifi-html-server/blob/main/src/wifi.rs) is almost the same as the one in the issue description)_.

It is possible, I might be missing something trivial, I apologize if that happens to be the case.

```bash
     Running `probe-rs run --chip=esp32c3 --preverify --always-print-stacktrace --no-location /Users
/vasanth/target/riscv32imc-unknown-none-elf/release/intro`
    Verifying ✔ 100% [####################] 339.60 KiB @  71.53 KiB/s (took 5s)
     Finished in 5.03s
[INFO ] Embassy initialized!
[INFO ] IPv4: DOWN
[INFO ] Waiting for link to be up
[INFO ] Connecting...
[ERROR] Connect failed: Disconnected(DisconnectedInfo { ssid: Ssid { ssid: [114, 101, 97, 108, 109,
101, 32, 80, 52, 120, 32, 53, 71, 32, 102, 56, 106, 113, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
len: 18 }, bssid: [194, 83, 13, 196, 223, 18], reason: AuthenticationExpired, rssi: -48 })
[INFO ] Connecting...
[ERROR] Connect failed: Disconnected(DisconnectedInfo { ssid: Ssid { ssid: [114, 101, 97, 108, 109,
101, 32, 80, 52, 120, 32, 53, 71, 32, 102, 56, 106, 113, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
len: 18 }, bssid: [0, 0, 0, 0, 0, 0], reason: ConnectionFailed, rssi: -128 })
[INFO ] Connecting...
[ERROR] Connect failed: Disconnected(DisconnectedInfo { ssid: Ssid { ssid: [114, 101, 97, 108, 109,
101, 32, 80, 52, 120, 32, 53, 71, 32, 102, 56, 106, 113, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
len: 18 }, bssid: [194, 83, 13, 196, 223, 18], reason: AuthenticationExpired, rssi: -52 })
```

--- a/esp-radio/src/wifi/mod.rs
+++ b/esp-radio/src/wifi/mod.rs
@@ -44,8 +44,8 @@
 //!
 //! If WiFi connections fail, especially on boards with a small or low-quality antenna or when the
 //! device is placed inside an enclosure, the TX power may need to be increased. The default
-//! maximum TX power is 20 (5dBm) on the 0.25dBm scale used by [`WifiController::set_max_tx_power`].
-//! Try increasing it if connections are unreliable.
+//! maximum TX power is 20 (5dBm) on the 0.25dBm scale used by `WifiController::set_max_tx_power`
+//! (requires the `unstable` feature). Try increasing it if connections are unreliable.
 
 use alloc::{borrow::ToOwned, collections::vec_deque::VecDeque, str, vec::Vec};
 use core::{

--- a/esp-radio/src/wifi/mod.rs
+++ b/esp-radio/src/wifi/mod.rs
@@ -42,10 +42,11 @@
 //!
 //! ### Connection failures on boards with a weak antenna or in a poor RF environment
 //!
-//! If WiFi connections fail, especially on boards with a small or low-quality antenna or when the
-//! device is placed inside an enclosure, the TX power may need to be increased. The default
-//! maximum TX power is 20 (5dBm) on the 0.25dBm scale used by `WifiController::set_max_tx_power`
-//! (requires the `unstable` feature). Try increasing it if connections are unreliable.
+//! The default maximum TX power is 20 (5dBm) on the 0.25dBm scale. The optimal value is
+//! board-dependent: if connections are unreliable, try adjusting it via
+//! `WifiController::set_max_tx_power` (requires the `unstable` feature) using a value in the
+//! range [8, 84]. Note that values above roughly 65 (~16dBm) have been reported to cause
+//! authentication failures on some hardware, so setting it to the maximum is not always better.
 
 use alloc::{borrow::ToOwned, collections::vec_deque::VecDeque, str, vec::Vec};
 use core::{
@@ -2090,9 +2091,10 @@ impl CountryInfo {
             // TODO: these may be valid defaults, but they should be configurable.
             schan: 1,
             nchan: 13,
-            // 20 on the 0.25dBm scale (5dBm). Can be overridden via
-            // WifiController::set_max_tx_power. See https://github.com/esp-rs/esp-hal/issues/5631#issuecomment-4634839995
-            max_tx_power: 20,
+            // This field is output-only: esp_wifi_set_country ignores it. The actual TX power
+            // is controlled exclusively via esp_wifi_set_max_tx_power after WiFi start.
+            // See: https://github.com/espressif/esp-idf/blob/20f5e18/components/esp_wifi/include/esp_wifi_types.h#L46
+            max_tx_power: 0,
             policy: wifi_country_policy_t_WIFI_COUNTRY_POLICY_MANUAL,
 
             #[cfg(wifi_has_5g)]
@@ -2394,6 +2396,9 @@ impl<'d> WifiController<'d> {
         controller.set_power_saving(PowerSaveMode::default())?;
 
         controller.set_config(&config.initial_config)?;
+
+        // Set a default TX power
+        esp_wifi_result!(unsafe { esp_wifi_set_max_tx_power(20) })?;
 
         Ok(controller)
     }
@@ -2813,10 +2818,9 @@ ignored."
     /// Set maximum transmitting power after WiFi start.
     ///
     /// Power unit is 0.25dBm, range is [8, 84] corresponding to 2dBm - 20dBm. The default is
-    /// 20 (5dBm). If WiFi connections are unreliable or you are experiencing problems connecting to
-    /// the AP, especially on boards with a weak or low-quality antenna, try increasing this
-    /// value. See the [module-level troubleshooting section](self#troubleshooting) for more
-    /// details.
+    /// 20 (5dBm). Values above roughly 65 (~16dBm) have been reported to cause authentication
+    /// failures on some hardware. See the
+    /// [module-level troubleshooting section](self#troubleshooting) for details.
     #[instability::unstable]
     pub fn set_max_tx_power(&mut self, power: i8) -> Result<(), WifiError> {
         esp_wifi_result!(unsafe { esp_wifi_set_max_tx_power(power) })

--- a/esp-radio/src/wifi/mod.rs
+++ b/esp-radio/src/wifi/mod.rs
@@ -37,6 +37,15 @@
 //!
 //! Please note that the configuration keys are usually named slightly different and not all
 //! configuration keys apply.
+//!
+//! ## Troubleshooting
+//!
+//! ### Connection failures on boards with a weak antenna or in a poor RF environment
+//!
+//! If WiFi connections fail, especially on boards with a small or low-quality antenna or when the
+//! device is placed inside an enclosure, the TX power may need to be increased. The default
+//! maximum TX power is 20 (5dBm) on the 0.25dBm scale used by [`WifiController::set_max_tx_power`].
+//! Try increasing it if connections are unreliable.
 
 use alloc::{borrow::ToOwned, collections::vec_deque::VecDeque, str, vec::Vec};
 use core::{
@@ -2081,6 +2090,8 @@ impl CountryInfo {
             // TODO: these may be valid defaults, but they should be configurable.
             schan: 1,
             nchan: 13,
+            // 20 on the 0.25dBm scale (5dBm). Can be overridden via
+            // WifiController::set_max_tx_power. See https://github.com/esp-rs/esp-hal/issues/5631#issuecomment-4634839995
             max_tx_power: 20,
             policy: wifi_country_policy_t_WIFI_COUNTRY_POLICY_MANUAL,
 
@@ -2801,7 +2812,11 @@ ignored."
 
     /// Set maximum transmitting power after WiFi start.
     ///
-    /// Power unit is 0.25dBm, range is [8, 84] corresponding to 2dBm - 20dBm.
+    /// Power unit is 0.25dBm, range is [8, 84] corresponding to 2dBm - 20dBm. The default is
+    /// 20 (5dBm). If WiFi connections are unreliable or you are experiencing problems connecting to
+    /// the AP, especially on boards with a weak or low-quality antenna, try increasing this
+    /// value. See the [module-level troubleshooting section](self#troubleshooting) for more
+    /// details.
     #[instability::unstable]
     pub fn set_max_tx_power(&mut self, power: i8) -> Result<(), WifiError> {
         esp_wifi_result!(unsafe { esp_wifi_set_max_tx_power(power) })

